ksql-engine/src/main/java/io/confluent/ksql/engine/TopicAccessValidator.java

```diff
@@ -0,0 +1,169 @@
+   *
+   * @param statement The statement to verify for permissions.
+   * @throws KsqlTopicAccessException If a topic does not exist or does not have permissions.
+   */
+  public void verifyStatementPermissions(final Statement statement)
+      throws KsqlTopicAccessException {
+    if (statement instanceof CreateSource) {
+      verifyCreateSourcePermissions((CreateSource) statement);
+    } else if (statement instanceof InsertInto) {
+      verifyInsertIntoPermissions((InsertInto) statement);
+    } else if (statement instanceof InsertValues) {
+      verifyInsertValuesPermissions((InsertValues) statement);
+    } else if (statement instanceof CreateAsSelect) {
+      verifyCreateAsSelectPermissions((CreateAsSelect) statement);
+    }
+  }
+
+  private void verifyCreateAsSelectPermissions(final CreateAsSelect createAsSelect) {
+    /*
+     * Check permissions for CREATE STREAM and CREATE TABLE statements that have
+     * a persistent query in it.
+     *
+     * This only validates the user can CREATE and WRITE the target topic. The persistent query
+     * should be validated by verifyQuerySourcesPermissions() when the logical plan is
+     * created (see EngineExecutor.execute)
+     */
+
+    final String kafkaTopic = createAsSelect.getName().getSuffix();
+    checkAccess(kafkaTopic, Sets.newHashSet(AclOperation.CREATE, AclOperation.WRITE));
+  }
+
+  private void verifyInsertIntoPermissions(final InsertInto insertInto) {
+    /*
+     * Check permissions for INSERT INTO statements that have a persistent query in it.
+     *
+     * This only validates the user can WRITE the target topic. The persistent query
+     * should be validated by verifyQuerySourcesPermissions() when the logical plan is
+     * created (see EngineExecutor.execute)
+     */
+
+    final String kafkaTopic = insertInto.getTarget().getSuffix();
+    checkAccess(kafkaTopic, Collections.singleton(AclOperation.WRITE));
+  }
+
+  private void verifyInsertValuesPermissions(final InsertValues insertValues) {
+    /*
+     * Check permissions for INSERT INTO VALUES statements that do not have
+     * a persistent query in it.
+     */
+
+    final String kafkaTopic = insertValues.getTarget().getSuffix();
+    checkAccess(kafkaTopic, Collections.singleton(AclOperation.WRITE));
+  }
+
+  private void verifyCreateSourcePermissions(final CreateSource createSource) {
+    /*
+     * Check permissions for CREATE STREAM and CREATE TABLE statements that do not have
+     * a persistent query in it. These statements come the WITH kafka_topic property.
+     */
+    final String kafkaTopic = createSource.getProperties().getKafkaTopic();
+    checkAccess(kafkaTopic, Collections.singleton(AclOperation.READ));
+  }
+
+  /**
+   * Walks through the PlanNode to find all the data sources of the query node.
+   */
+  private void collectKafkaTopics(final Set<String> kafkaTopics, final PlanNode node) {
```

Comment:
this feels like a method that should be part of `PlanNode#getAllSourceKafkaTopics`. That way this class doesn't need to know anything about logical plans and will just accept a `Set<String>`